Fix #5052: Create one-off schema migration jobs for topics, skills, questions and stories

nalinbhardwaj opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include 
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue 
  - when this PR is merged.
  -->
Creates one-off schema migration jobs for topics, skills, questions and stories, along with appropriate modifications to `services` and `domain` to support schema migration. Adds very basic tests for each of the jobs as well.

Fixes #5052.
Fixes nits in `collection_jobs_one_off_test.py` as well.

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes.
- [x] The PR explanation includes the words "Fixes #bugnum: ...".
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

nalinbhardwaj committed:
collection_jobs_one_off_test : fix nits

- Remove extraneous definition
- Fix comment typo

nalinbhardwaj committed:
skill_jobs_one_off : Create schema migration jobs for skill

Adds basic tests as well.

nalinbhardwaj committed:
story_jobs_one_off : Create schema migration jobs for story

Adds basic tests as well.

nalinbhardwaj committed:
topic_jobs_one_off : Create schema migration jobs for topic

Adds basic tests as well.

nalinbhardwaj committed:
question_jobs_one_off : Create schema migration jobs for question

Adds basic tests as well.

seanlip commented on core/domain/question_jobs_one_off_test.py:
```diff
@@ -0,0 +1,98 @@
+(question_models,) = models.Registry.import_models([models.NAMES.question])
+
+
+class QuestionMigrationJobTest(test_utils.GenericTestBase):
```
Can you also find a way to add tests where a question is created with schema version v1 and we check, on-load, that it is the latest schema version? That way we can make sure that every conversion function works going forward.

Note that questions don't have YAML representations so we may need to find another way, perhaps we could use dicts and from_dict() or similar.

nalinbhardwaj replied:
I'm not sure if I understand this correctly. Firstly, you mean v25, not v1 right?
Secondly, for this and the general "tests that ensure v1 is converted to vN", since these structures haven't been released, this test won't actually test any conversion functions(which don't exist either?). Is it intended for future use? If so, how would this test create a old version model(using `from_dict` etc., as you suggest) when the functions then would just generate a model with the newest version at that time? Besides, are the original and expected models just hardcoded(or generated somehow?)?

seanlip replied:
Great questions, thanks!

Yes, I mean v25, sorry.

Also, yes, it is intended for future use. If you look at the exploration migration tests, we have a bunch of hardcoded YAML representing old versions; we should do something similar here. I'm not sure if from_dict() with an old schema version would work; my guess is that it might not (as you suggest). So perhaps an alternative way is to handcraft a storage model instance using hardcoded dicts, and then try migrating that instead.

nalinbhardwaj replied:
Fixed.

seanlip commented on core/domain/question_services.py:
```diff
@@ -137,20 +173,44 @@ def delete_question(
 
-def get_question_from_model(question_model):
+def get_question_from_model(question_model, run_conversion=True):
     """Returns domain object repersenting the given question model.
```
Not your fault, but please could you change repersenting to representing here while you're at it?

nalinbhardwaj replied:
fixed

seanlip commented on core/domain/question_services.py:
```diff
@@ -137,20 +173,44 @@ def delete_question(
 
 
-def get_question_from_model(question_model):
+def get_question_from_model(question_model, run_conversion=True):
```
Should the new option come with tests? I assume that's what the optional arg is for...

nalinbhardwaj replied:
The docstring does imply that the option is for testing, but a [quick search](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/oppia/oppia%24+run_conversion%3DFalse) shows that right now there are no specific tests(in the existing `get_*_from_model` functions) for the option.

seanlip replied:
I think we can remove the option then, in all places.

nalinbhardwaj replied:
removed

seanlip commented on core/domain/skill_jobs_one_off.py:
```diff
@@ -0,0 +1,102 @@
+            commit_cmds.append(skill_domain.SkillChange({
+                'cmd': skill_domain.CMD_MIGRATE_CONTENTS_SCHEMA_TO_LATEST_VERSION, # pylint: disable=line-too-long
+                'from_version': item.skill_contents_schema_version,
+                'to_version': str(
```
Why do we stringify this, here and elsewhere? /cc @aks681

aks681 replied:
I am not sure. As far as I can tell, the value is not directly used anywhere when `update_skill` and hence is only used for logging purposes, when model.commit() is called with this changelist. So, I think it shouldn't matter whether it is int ot string, since the `from_version` above is an integer.
@nalinbhardwaj Can you confirm this?

nalinbhardwaj replied:
Seems this started in https://github.com/oppia/oppia/commit/7aa7a6b449a71301153d006924f696cf4292bab3#diff-e9c1ba30dea3ada6c70e548655621ea2R129 when both `from_version` and `to_version` were `str`s but I can't find much rationale in it for this particular decision. As far as I can tell, it is indeed only used for logging purposes. Removing it doesn't seem to break anything as well, so I'm going to remove str.

seanlip commented on core/domain/skill_jobs_one_off_test.py:
```diff
@@ -0,0 +1,103 @@
+        self.signup(self.ALBERT_EMAIL, self.ALBERT_NAME)
+        self.process_and_flush_pending_tasks()
+
+    def test_migration_job_does_not_convert_up_to_date_skill(self):
```
Ditto re tests that ensure v1 is converted to vN.

Also try and find ways to test situations where (i) both schemas have updates, (ii) one schema has updates but the other does not.

nalinbhardwaj replied:
Done. Note that I haven't added actual tests that do (ii) yet, but with the `test_util` function in place, it is just a single call change in the existing tests to implement this. I think they should only be added depending on the future version changes.

seanlip commented on core/domain/story_jobs_one_off_test.py:
```diff
@@ -0,0 +1,97 @@
+(story_models,) = models.Registry.import_models([models.NAMES.story])
+
+
+class StoryMigrationJobTest(test_utils.GenericTestBase):
```
Ditto.

nalinbhardwaj replied:
fixed

seanlip commented on core/domain/topic_jobs_one_off.py:
```diff
@@ -0,0 +1,88 @@
+(topic_models,) = models.Registry.import_models([models.NAMES.topic])
+
+
+class TopicMigrationJob(jobs.BaseMapReduceOneOffJobManager):
```
Here and elsewhere, let's maybe name these to end with *OneOffJob, for clarity.

Also, I noticed that these jobs don't seem to be in the job registry. Did you try running them manually?

nalinbhardwaj replied:
> Here and elsewhere, let's maybe name these to end with *OneOffJob, for clarity.

Sure, but that would break convention from `collection_jobs_one_off.CollectionMigrationJob` and `exp_jobs_one_off.ImageDataMigrationJob`. Should they be changed as well?

> Also, I noticed that these jobs don't seem to be in the job registry. Did you try running them manually?

Yes, but I omitted putting the jobs in the job registry before there are multiple versions of models. As far as I understand, they are unreleased so it doesn't make sense to have these in the job registry right now? Or should they be added regardless?

seanlip replied:
Oh, that's a good point. I think, let's put them in if running them is a no-op, but guard the map() with a check on feconf.ENABLE_NEW_STRUCTURES. Then, when that flag is removed, everything will behave correctly post-removal.

nalinbhardwaj replied:
> Oh, that's a good point. I think, let's put them in if running them is a no-op, but guard the map() with a check on feconf.ENABLE_NEW_STRUCTURES. Then, when that flag is removed, everything will behave correctly post-removal.

fixed

> Sure, but that would break convention from collection_jobs_one_off.CollectionMigrationJob and exp_jobs_one_off.ImageDataMigrationJob. Should they be changed as well?

can you confirm this as well?

nalinbhardwaj replied:
renamed.

seanlip reviewed (commented):
Thanks @nalinbhardwaj! I took a pass.

aks681 commented on core/domain/skill_domain.py:
```diff
@@ -124,6 +125,12 @@ def __init__(self, change_dict):
             self.old_value = copy.deepcopy(change_dict['old_value'])
         elif self.cmd == CMD_CREATE_NEW:
             return
+        elif self.cmd == CMD_MIGRATE_CONTENTS_SCHEMA_TO_LATEST_VERSION:
```
Add these also to the docstring for this function.

nalinbhardwaj replied:
fixed

aks681 commented on core/domain/question_domain.py:
```diff
@@ -79,32 +86,31 @@ def __init__(self, change_dict):
             # Note that change_dict['skill_id'] may be None if this change is
             # being done in the context of a suggestion.
             self.skill_id = change_dict['skill_id']
+        elif self.cmd == CMD_MIGRATE_STATE_SCHEMA_TO_LATEST_VERSION:
```
Ditto docstring.

nalinbhardwaj replied:
fixed

aks681 commented on core/domain/topic_domain.py:
```diff
@@ -118,6 +121,9 @@ def __init__(self, change_dict):
             self.property_name = change_dict['property_name']
             self.new_value = copy.deepcopy(change_dict['new_value'])
             self.old_value = copy.deepcopy(change_dict['old_value'])
+        elif self.cmd == CMD_MIGRATE_SUBTOPIC_SCHEMA_TO_LATEST_VERSION:
```
Ditto docstring.

nalinbhardwaj replied:
fixed

aks681 reviewed (commented):
Did a pass. Looks good overall!

nalinbhardwaj committed:
fixup! skill_jobs_one_off : Create schema migration jobs for skill

nalinbhardwaj committed:
fixup! story_jobs_one_off : Create schema migration jobs for story

nalinbhardwaj committed:
fixup! topic_jobs_one_off : Create schema migration jobs for topic

nalinbhardwaj committed:
fixup! question_jobs_one_off : Create schema migration jobs for question

nalinbhardwaj committed:
*_services: remove run_conversion from get_*_from_model

nalinbhardwaj committed:
job_registry: Add jobs to jobs_registry.

Guards jobs using feconf.ENABLE_NEW_STRUCTURES as well.
Modifies tests accordingly.

nalinbhardwaj committed:
commit_cmd from_version and to_version to int

Changes from_version and to_version to be int, not str, in existing one off jobs.

seanlip commented on core/domain/collection_jobs_one_off.py:
```diff
@@ -71,8 +71,7 @@ def map(item):
                 'from_version': item.schema_version,
-                'to_version': str(
-                    feconf.CURRENT_COLLECTION_SCHEMA_VERSION)
+                'to_version': feconf.CURRENT_COLLECTION_SCHEMA_VERSION
```
Thinking about it -- let's leave this as it was before, but fix the newer ones, and above this add a comment "This really should be int, but left as str to conform with legacy data" or similar. I think new models should be handled correctly; however, for existing models, we don't want our datastore to have some existing model instances follow one convention and new models follow another. The aim of the comment is to flag why the difference is there so that future devs don't unwittingly change the code.

(Does that sound good?)

nalinbhardwaj replied:
Sounds good, although I think it's safe to change these, but it's not worth breaking something due to something so minor. I've made the comments.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -246,9 +246,8 @@ def map(item):
-                'from_version': str(item.states_schema_version),
-                'to_version': str(
-                    feconf.CURRENT_STATES_SCHEMA_VERSION)
+                'from_version': item.states_schema_version,
```
Ditto, here and next line.

nalinbhardwaj replied:
Done.

seanlip commented on core/domain/question_jobs_one_off.py:
```diff
@@ -0,0 +1,93 @@
+        if not feconf.ENABLE_NEW_STRUCTURES:
+            yield (
+                QuestionMigrationJob._DISABLED_KEY,
+                'Encountered disabled structure.')
```
'New structures framework is disabled.'

nalinbhardwaj replied:
fixed.

seanlip commented on core/domain/question_jobs_one_off.py:
```diff
@@ -0,0 +1,93 @@
+        if item.deleted:
+            yield (
+                QuestionMigrationJob._DELETED_KEY,
+                'Encountered deleted question.')
```
Yield 1 here. Aggregate sums in reduce() and yield just the sum of occurrences, since this is an expected behavior (i.e. some instances will, in general, have been deleted, so there's no need to yield instance-specific info for them).

In general, for expected behavior record-keeping, yield aggregate counts. For cases where we are trying to detect errors, yield specific IDs so that we can debug. (OK to apply this logic retrospectively to existing jobs, if needed.)

nalinbhardwaj replied:
Should this be changed in [`CollectionMigrationJob`](https://github.com/oppia/oppia/blob/develop/core/domain/collection_jobs_one_off.py#L53) as well?

seanlip replied:
Yes, let's do that.

nalinbhardwaj replied:
done.

seanlip reviewed (commented):
Hi @nalinbhardwaj, thanks! I was only able to take a pass at the first few files, but comments there should apply to the other models as well.

> Sure, but that would break convention from collection_jobs_one_off.CollectionMigrationJob and exp_jobs_one_off.ImageDataMigrationJob. Should they be changed as well?
> can you confirm this as well?

Yep let's change them. Thanks! (And sorry for missing replying to this earlier.)

nalinbhardwaj committed:
Revert "commit_cmd from_version and to_version to int"

This reverts commit ad8e497473443bd386e1431b70be7f4e86f248e4.

nalinbhardwaj committed:
commit_cmd: add note for version fields

nalinbhardwaj committed:
fixup! job_registry: Add jobs to jobs_registry.

nalinbhardwaj committed:
Use constants.ENABLE_NEW_STRUCTURES

Changes due to rebase over `8f3d11c`.

nalinbhardwaj committed:
Change how deleted items are handled

https://github.com/oppia/oppia/pull/5660/files#r220786659

nalinbhardwaj committed:
rename jobs as *OneOffJob

nalinbhardwaj committed:
create test_utils functions and tests for updating old schemas

nalinbhardwaj commented:
@seanlip / @aks681 : Sorry for the delays on this PR, this is ready for review again.

I should note that I'm not very confident if the way I implemented the `test_util`functions is the best way, since it feels very repetitive. Let me know if you have better ideas for their implementations. I think committing them  to the datastore manually is the right way, since it is hard to tell how `from_dict` etc. can change with future schema versions.

seanlip commented on core/domain/collection_jobs_one_off.py:
```diff
@@ -79,9 +80,16 @@ def map(item):
     @staticmethod
     def reduce(key, values):
-        yield (key, values)
+        if key == CollectionMigrationOneOffJob._DELETED_KEY:
```
Typically we use ast.literal_eval() to handle 'values' first, and then you don't need to do any further casting. Consider doing that instead?

(Ditto in other jobs, if applicable.)

nalinbhardwaj replied:
done.

seanlip commented on core/domain/collection_jobs_one_off.py:
```diff
@@ -79,9 +80,16 @@ def map(item):
                     feconf.CURRENT_COLLECTION_SCHEMA_VERSION))
             yield (
-                CollectionMigrationJob._MIGRATED_KEY,
+                CollectionMigrationOneOffJob._MIGRATED_KEY,
```
I suggest handling this in the same way as the deleted case -- use counts.

(Ditto in other jobs, if applicable.)

nalinbhardwaj replied:
This was done too, sorry I forgot to mention before.

seanlip commented on core/domain/question_jobs_one_off.py:
```diff
@@ -0,0 +1,99 @@
+    def map(item):
+        if not constants.ENABLE_NEW_STRUCTURES:
+            yield (
+                QuestionMigrationOneOffJob._DISABLED_KEY,
```
Handle this in the same way as "deleted", since this is an expected thing.

Only yield IDs for unexpected error cases (so that we can debug).

(Ditto elsewhere.)

nalinbhardwaj replied:
I thought about this, but it doesn't seem very useful since when these structures are released, these guards should simply be removed(and before that, they would not trigger since there exist no datastores). Doing this would just increase the amount of code we need to remove on a later date. Is that intended or am I misunderstanding something?

seanlip replied:
That's a very fair point. I was going according to the general principle of "aggregate if expected in practice", but I realize from your comment that this is not expected in practice! So on balance I think that what you have now is fine, and I like your reasoning. Thank you for clarifying!

seanlip reviewed (commented):
Hi @nalinbhardwaj, thanks! I took a look and found some things to tidy up in the jobs (which are otherwise really nicely done).

@aks681 could you PTAL too?

nalinbhardwaj committed:
Use ast.literal_eval in reduce and change how migrated items are handled.

seanlip commented:
Hmm, I don't see your changes in the "Files Changed" tab. Did you forget to
push them?

(In general, always check that tab before requesting re-review. Thanks!)

>

nalinbhardwaj commented:
@seanlip : No, I did push them.
 https://github.com/oppia/oppia/pull/5660/commits/eab933072e14c03cfc19c2fe5750e281e74912e3 has the changes for the aggregates of the migrated items(its not a new commit, I just forgot to mention that I made it ~15 hours ago)

seanlip commented:
Weird. Does it show up for you in the Files Changed tab on GitHub? It
doesn't for me, and that's what I use as a basis for review.

>

nalinbhardwaj commented:
@seanlip : yes, it does show up for me. Maybe it's not showing for you because I rebased and resolved conflicts earlier yesterday. Anyways, the latest commit contains those changes.

seanlip commented:
Nalin, is this the page you are looking at?
https://github.com/oppia/oppia/pull/5660/files

@aks681 do you see the changes in response to the latest round of comments?

(Also in general I suggest not rebasing - follow the process on the dev
wiki. It can lead to issues, eg with not being able to easily review the
diff from the previous review. Thanks!)

nalinbhardwaj commented:
@seanlip : yes, thats the page. See https://github.com/oppia/oppia/commit/eab933072e14c03cfc19c2fe5750e281e74912e3 from that page. It should show the changes...
I’ll keep that in mind for future. Sorry.

seanlip commented:
Ah, I was looking at the collection jobs file. It looks like that hasn't
changed and is not included in the commit you pointed out?

nalinbhardwaj committed:
fix migrated item handling for collections

nalinbhardwaj commented:
@seanlip : Ohh, yeah you're right I missed that file in particular. My bad.

Fixed now. Sorry.

seanlip commented on core/domain/collection_jobs_one_off.py:
```diff
@@ -78,10 +80,15 @@ def map(item):
-        yield (key, values)
+        if key == CollectionMigrationOneOffJob._DELETED_KEY:
+            yield (key, ['Encountered %d deleted collections.' % (
+                sum(ast.literal_eval(v) for v in values))])
```
Could you please add backend tests, here and elsewhere, to cover this new code?

nalinbhardwaj replied:
I'm not sure if I understand what the test is expected to be. Can you point me to any other similar tests in the repo?

seanlip replied:
It would be similar to existing tests in collection_jobs_one_off_test.py (and other one-off test files). In this case, you could set up a collection, delete it, run the job, and then check that the output of the job is as expected. (Note that the core job classes are defined in core/jobs.py.)

nalinbhardwaj replied:
fixed.

seanlip commented on core/domain/question_jobs_one_off.py:
```diff
@@ -0,0 +1,98 @@
+    @staticmethod
+    def map(item):
+        if not constants.ENABLE_NEW_STRUCTURES:
+            yield (
```
Actually, thinking about this more -- shouldn't this just "return"? Otherwise you'll get a list of repeated "New structures framework is disabled." values and I'm not sure that that's useful.

(Ditto elsewhere. Make sure to verify behaviour with backend tests.)

nalinbhardwaj committed:
update tests to test job outputs and update jobs to remove disabled key

seanlip commented on core/tests/test_utils.py:
```diff
@@ -53,7 +53,9 @@
 import webtest
 
-(exp_models,) = models.Registry.import_models([models.NAMES.exploration])
+(exp_models, question_models, skill_models, story_models, topic_models,) = models.Registry.import_models([ # pylint: disable=line-too-long
```
You can break this line:

```
(exp_models, question_models, skill_models, story_models, topic_models,) = (
    models.Registry.import_models([
        models.NAMES.exploration, models.NAMES.question, models.NAMES.skill,
        models.NAMES.story, models.NAMES.topic]))
```

nalinbhardwaj replied:
fixed :)

seanlip reviewed (approved):
Hi @nalinbhardwaj, this looks great! Just one more small comment, then it can be merged. Thanks!

seanlip commented:
Also, I restarted the embedding e2e test. Let's see if it works and if the earlier issue was just a flake.

nalinbhardwaj committed:
break lines

apb7 commented:
@seanlip, @nalinbhardwaj : This seems to be a flake and occurs randomly infrequently probably due to Travis not being able to read from cache. I am going ahead and deleting the cache for this branch. Thanks!

seanlip commented:
It works! Thanks @apb7. Merging!
